nexus_motion_planner/src/motion_plan_cache.cpp

```diff
@@ -0,0 +1,935 @@
+        "Skipping insert: Could not construct insert metadata.");
+      return false;
+    }
+
+    RCLCPP_INFO(
+      node_->get_logger(),
+      "Inserting plan: New plan execution_time (%es) "
+      "is better than best plan's execution_time (%es)",
+      execution_time_s, best_execution_time_seen);
+
+    coll.insert(plan, insert_metadata);
+    return true;
+  }
+
+  RCLCPP_INFO(
+    node_->get_logger(),
+    "Skipping insert: New plan execution_time (%es) "
+    "is worse than best plan's execution_time (%es)",
+    execution_time_s, best_execution_time_seen);
+  return false;
+}
+
+// MOTION PLAN CACHING: QUERY CONSTRUCTION
+bool
+MotionPlanCache::extract_and_append_plan_start_to_query(
+  Query& query,
+  const moveit::planning_interface::MoveGroupInterface& move_group,
+  const moveit_msgs::msg::MotionPlanRequest& plan_request,
+  double match_tolerance)
+{
+  match_tolerance += exact_match_precision_;
+
+  // Make ignored members explicit
+  if (!plan_request.start_state.multi_dof_joint_state.joint_names.empty())
+  {
+    RCLCPP_WARN(
+      node_->get_logger(),
+      "Ignoring start_state.multi_dof_joint_states: Not supported.");
+  }
+  if (!plan_request.start_state.attached_collision_objects.empty())
+  {
+    RCLCPP_WARN(
+      node_->get_logger(),
+      "Ignoring start_state.attached_collision_objects: Not supported.");
+  }
+
+  // auto original = *query;  // Copy not supported.
+
+  query.append("group_name", plan_request.group_name);
+
+  // Workspace params
+  // Match anything within our specified workspace limits.
+  query.append(
+    "workspace_parameters.header.frame_id",
+    plan_request.workspace_parameters.header.frame_id);
+  query.appendGTE(
+    "workspace_parameters.min_corner.x",
+    plan_request.workspace_parameters.min_corner.x);
+  query.appendGTE(
+    "workspace_parameters.min_corner.y",
+    plan_request.workspace_parameters.min_corner.y);
+  query.appendGTE(
+    "workspace_parameters.min_corner.z",
+    plan_request.workspace_parameters.min_corner.z);
+  query.appendLTE(
+    "workspace_parameters.max_corner.x",
+    plan_request.workspace_parameters.max_corner.x);
+  query.appendLTE(
+    "workspace_parameters.max_corner.y",
+    plan_request.workspace_parameters.max_corner.y);
+  query.appendLTE(
+    "workspace_parameters.max_corner.z",
+    plan_request.workspace_parameters.max_corner.z);
+
+  // Joint state
+  //   Only accounts for joint_state position. Ignores velocity and effort.
+  if (plan_request.start_state.is_diff)
+  {
+    // If plan request states that the start_state is_diff, then we need to get
+    // the current state from the move_group.
+
+    // NOTE: methyldragon -
+    //   I think if is_diff is on, the joint states will not be populated in all
+    //   of our motion plan requests? If this isn't the case we might need to
+    //   apply the joint states as offsets as well.
+    moveit::core::RobotStatePtr current_state = move_group.getCurrentState();
```

Comment:
iiuc this can be flimsy, the state we retrieve here may not be the same as the state moveit used for planning. How much impact would it be if we drop support for `is_diff`?

Reply:
`is_diff` is the default way plans are made (if drawn from a start state from a TF frame.)

The `extract_and_append_plan_start_to_query` function is used for fetching plans, so we get the current state of the robot to pull up plans in the cache which are close-enough to the current state of the robot, wrt. **concrete** joint-state values.

This way, we avoid that flimsiness you mentioned.

Reply:
My concern is like so:

1. Motion plan server receives plan request with `is_diff`.
2. Pass request to moveit, moveit get the current robot state and does the planning.
3. Motion plan server receive the plan, get the current robot state again and use it to key the cache.
4. At step 2, we may get joint 1 = `0.5`, then at step 3, we get joint 1 = `0.501`. This can cause "cache poisoning" if the difference is too big.

It would be better if there is a way to get the exact state moveit used to do the planning, or is that infact what `move_group.getCurrentState()` is doing? I'm not experienced with moveit so I thought it is getting the robot state from `ros_control`.

Reply:
Hmm, I think getCurrentState() is getting it from the planning scene monitor, so most likely ros_control (unless the state was set manually otherwise)...

I don't think this is a problem we can solve at this level, and since we want to upstream this (so maintenance is up to the MoveIt maintainers), I don't think it's a good idea to drop support for `is_diff` too...

Ideally when it gets upstreamed, the caching should happen as part of the plan call within MoveIt itself.

I don't think this is a problem in our application, so I think we should TODO this for now.

bbc07cec581f5b127ec7f86227da69b078994a38